skxray/cdi.py

```diff
@@ -119,9 +125,339 @@ def convolution(array1, array2):
     `this issue on github <https://github.com/Nikea/scikit-xray/issues/258>`_
     for details.
     """
-    fft_norm = lambda x:  np.fft.fftshift(np.fft.fftn(x)) / np.sqrt(np.size(x))
-    fft_1 = fft_norm(array1)
-    fft_2 = fft_norm(array2)
-    return np.abs(np.fft.ifftshift(np.fft.ifftn(fft_1*fft_2)) *
-                  np.sqrt(np.size(array2)))
+    fft_1 = _fft_helper(array1) / np.sqrt(np.size(array1))
+    fft_2 = _fft_helper(array2) / np.sqrt(np.size(array2))
+    return np.abs(_ifft_helper(fft_1*fft_2) * np.sqrt(np.size(array2)))
+
+
+def pi_modulus(array_in, diff_array,
+               thresh_v=1e-12):
+    """
+    Transfer sample from real space to q space.
+    Use constraint based on diffraction pattern from experiments.
+
+    Parameters
+    ----------
+    array_in : array
+        reconstructed pattern in real space
+    diff_array : array
+        experimental data
+    thresh_v : float
+        add small value to avoid the case of dividing something by zero
+
+    Returns
+    -------
+    arr : array
+        updated pattern in q space
+    """
+    diff_tmp = _fft_helper(array_in) / np.sqrt(np.size(array_in))
+    index = np.where(diff_array > 0)
+    diff_tmp[index] = diff_array[index] * diff_tmp[index] / (np.abs(diff_tmp[index]) + thresh_v)
+
+    return _ifft_helper(diff_tmp) * np.sqrt(np.size(diff_array))
+
+
+def find_support(sample_obj,
+                 sw_sigma, sw_threshold):
+    """
+    Update sample area based on thresholds.
+
+    Parameters
+    ----------
+    sample_obj : array
+        sample for reconstruction
+    sw_sigma : float
+        sigma for gaussian in shrinkwrap method
+    sw_threshold : float
+        threshold used in shrinkwrap method
+
+    Returns
+    -------
+    new_sup : array
+        updated sample support
+    s_index : array
+        index for sample area
+    s_out_index : array
+        index not for sample area
+    """
+
+    sample_obj = np.abs(sample_obj)
+    gauss_fun = gauss(sample_obj.shape, sw_sigma)
```

Comment:
The support of the gaussian should be set by the sigma.  You only need support out to a couple of standard deviations.  Using a smaller kernel will speed up the computation.